test/e2e/federation-ingress.go

```diff
@@ -20,285 +20,128 @@ import (
-		Describe("Ingress creation", func() {
-			BeforeEach(func() {
-				framework.SkipUnlessFederated(f.Client)
-				// Placeholder
-			})
-
-			AfterEach(func() {
-				framework.SkipUnlessFederated(f.Client)
-				// Placeholder
-			})
+		conformanceTests = createComformanceTests(jig, ns)
 
-			It("should succeed", func() {
-				framework.SkipUnlessFederated(f.Client)
-				ingress := createIngressOrFail(f.FederationClientset_1_3, f.Namespace.Name)
-				By(fmt.Sprintf("Creation of ingress %q in namespace %q succeeded.  Deleting ingress.", ingress.Name, f.Namespace.Name))
-				// Cleanup
-				//TODO wait for ingress object in federation apiserver
-				err := f.FederationClientset_1_3.Ingress(f.Namespace.Name).Delete(ingress.Name, &api.DeleteOptions{})
-				framework.ExpectNoError(err, "Error deleting ingress %q in namespace %q", ingress.Name, ingress.Namespace)
-				By(fmt.Sprintf("Deletion of ingress %q in namespace %q succeeded.", ingress.Name, f.Namespace.Name))
-			})
-
-			It("should create matching ingresss in underlying clusters", func() {
-				framework.SkipUnlessFederated(f.Client)
-				ingress := createIngressOrFail(f.FederationClientset_1_3, f.Namespace.Name)
-				defer func() { // Cleanup
-					By(fmt.Sprintf("Deleting ingress %q in namespace %q", ingress.Name, f.Namespace.Name))
-					//TODO wait for ingress object in federation apiserver
-					err := f.FederationClientset_1_3.Ingress(f.Namespace.Name).Delete(ingress.Name, &api.DeleteOptions{})
-					framework.ExpectNoError(err, "Error deleting ingress %q in namespace %q", ingress.Name, f.Namespace.Name)
-				}()
-				waitForIngressShardsOrFail(f.Namespace.Name, ingress, clusters)
-			})
+		AfterEach(func() {
+			teardownClusters(clusters, f)
 		})
 
-		var _ = Describe("Ingress Provider Validation", func() {
-
-			var (
-				ingress *extensions.Ingress
-			)
+		Describe("Ingress creation", func() {
 
-			//TODO for ingress, the controller does only rely on pod status only, need add more objects creation and check
 			BeforeEach(func() {
 				framework.SkipUnlessFederated(f.Client)
-				createBackendPodsOrFail(clusters, f.Namespace.Name, FederatedIngressPodName)
-				ingress = createIngressOrFail(f.FederationClientset_1_3, f.Namespace.Name)
-				waitForIngressShardsOrFail(f.Namespace.Name, ingress, clusters)
+				framework.SkipUnlessProviderIs("gce", "gke")
+				By("Initializing gce controller")
+				gceController = &GCEIngressController{ns: ns, Project: framework.TestContext.CloudConfig.ProjectID, c: jig.client}
+				gceController.init()
 			})
 
-			//TODO clear all object being created in BeforeEach()
 			AfterEach(func() {
 				framework.SkipUnlessFederated(f.Client)
-				deleteBackendPodsOrFail(clusters, f.Namespace.Name)
-
-				if ingress != nil {
-					deleteIngressOrFail(f.FederationClientset_1_3, f.Namespace.Name, ingress.Name)
-					ingress = nil
-				} else {
-					By("No ingress to delete.  Ingress is nil")
+				if CurrentGinkgoTestDescription().Failed {
+					describeIng(ns)
```

Comment:
This is a function defined in https://github.com/kubernetes/kubernetes/blob/master/test/e2e/ingress.go#L190, which was extracted to ingress-utils.go, the only purpose is to dump ingress object info in case e2e test fail.